### DISCUSSION TITLE
failed to connect: ssl set client certificate failed

### DISCUSSION THREAD

**Discussion by zhao173205** [Kong/kong#12381] [Help]
kong 3.4.2
I have developed a proxy bidirectional authentication Lua plugin myself. When I upgraded from version 3.1.1 to 3.4.2, the plugin kept loading certificates with the error "failed to connect: ssl set client certificate failed"

[error] 1260#0: *2349 [kong] access.lua:108 [http-proxy] retries: 3, failed to connect: ssl set client certificate failed

Some of the codes are as follows
local http = require "resty.http"
local ssl = require "ngx.ssl"
local cjson = require "cjson.safe"
local tools_utils = require "kong.tools.utils"
local utils = require("kong.router.utils")
local sanitize_uri_postfix = utils.sanitize_uri_postfix
local strip_uri_args = utils.strip_uri_args
local get_service_info = utils.get_service_info
local get_upstream_uri_v0 = utils.get_upstream_uri_v0

_M = {}

function _M.execute(conf)
if conf.log_enable then
kong.log("ngx.ctx.balancer_data: ", cjson.encode(ngx.ctx.balancer_data))
kong.log("kong.router.get_route: ", cjson.encode(kong.router.get_route()))
kong.log("kong.router.get_service: ", cjson.encode(kong.router.get_service()))
end

local proxy_options = {
http_proxy = "http://" .. conf.host .. ":" .. conf.port,
https_proxy = "http://" .. conf.host .. ":" .. conf.port,
}
if conf.mtls_enable then
local f = assert(io.open(conf.ssl_client_cert))
local cert_data = f:read("*a")
f:close()
CERT, ERR = ssl.parse_pem_cert(cert_data)
if not CERT then
kong.log("failed to parse pem cert: ", ERR)
return
end
local f_key = assert(io.open(conf.ssl_client_priv_key))
local cert_key_data = f_key:read("*a")
f_key:close()

CERT_KEY, ERRR = ssl.parse_pem_priv_key(cert_key_data)
if not CERT_KEY then
  kong.log("failed to parse pem cert_key: ", ERRR)
  return
end
end

local connect_options = {
scheme = ngx.ctx.balancer_data.scheme,
host = ngx.ctx.balancer_data.host,
port = ngx.ctx.balancer_data.port,
ssl_verify = true,
ssl_server_name = ngx.ctx.balancer_data.host,
proxy_opts = proxy_options,
}
if conf.log_enable then
kong.log("connect_options: ", cjson.encode(connect_options))
end

local request_params = {
version = kong.request.get_http_version(),
method = kong.request.get_method(),
path = kong.request.get_raw_path(),
query = kong.request.get_raw_query(),
headers = kong.request.get_headers(),
body = kong.request.get_raw_body(),
}
request_params.headers["remoteip"] = nil
request_params.headers["x-forwarded-for"] = nil
request_params.headers["host"] = ngx.ctx.balancer_data.host
-- get upstream uri start
local req_uri = kong.request.get_raw_path()
local matched_path = kong.request.get_forwarded_prefix()
local service = kong.router.get_service()
local matched_route = kong.router.get_route()
local service_protocol, _, --service_type
service_host, service_port,
service_hostname_type, service_path = get_service_info(service)
local request_prefix = matched_route.strip_path and matched_path or nil
local request_postfix = request_prefix and req_uri:sub(#matched_path + 1) or req_uri:sub(2, -1)
request_postfix = sanitize_uri_postfix(request_postfix) or ""
local upstream_base = service_path or "/"
local upstream_uri = get_upstream_uri_v0(matched_route, request_postfix, req_uri, upstream_base)
-- get upstream uri end
request_params.path = upstream_uri
if conf.log_enable then
kong.log("request_params: ", cjson.encode(request_params))
end

local max_retries = ngx.ctx.balancer_data.retries
local success = false
local errinfo = ""
local retries = 0
local res_status, res_body, res_headers

while not success and retries < max_retries do
local httpc = http.new()
httpc:set_timeouts(ngx.ctx.balancer_data.connect_timeout, ngx.ctx.balancer_data.send_timeout,
ngx.ctx.balancer_data.read_timeout)
if conf.mtls_enable then
-- httpc:setclientcert(CERT, CERT_KEY)
httpc.sock:setclientcert(CERT, CERT_KEY)
print("Debug: connect_options CERT: ",tostring(CERT))
print("Debug: connect_options CERT_KEY: ",tostring(CERT_KEY))
end
-- httpc:set_proxy_options(proxy_options)
while true do
local ok, err = httpc:connect(connect_options)
if not ok then
retries = retries + 1
errinfo = "retries: " .. retries .. ", failed to connect: " .. err
kong.log.err(errinfo)
break
end
local res, err = httpc:request(request_params)
if not res then
retries = retries + 1
errinfo = "retries: " .. retries .. ", failed to request: " .. err
kong.log.err(errinfo)
break
end
local response_info = {
status = res.status,
reason = res.reason,
headers = res.headers,
has_body = res.has_body,
}
if conf.log_enable then
kong.log("response info: ", cjson.encode(response_info))
end
local body, err = res:read_body()
if conf.log_enable then
if res.headers["Content-Encoding"] == "gzip" then
kong.log("response body: ", assert(tools_utils.inflate_gzip(body)))
else
kong.log("response body: ", body)
end
end
res_status = res.status
res_body = body
res_headers = res.headers
success = true
break
end
httpc:set_keepalive()
end

if not success then
-- operation failed after maximum retries
return kong.response.exit(500, { message = errinfo })
end

res_headers["Transfer-Encoding"] = nil
res_headers["transfer-encoding"] = nil
kong.response.exit(res_status, res_body, res_headers)
end

return _M

**Comment by Water-Melon** [comment]
It seems to be related to the OpenSSL version upgrade.

Kong3.1 uses OpenSSL 1.1.1s, while kong 3.4 uses OpenSSL 3.1.4. TLSv1.1 is not supported by default in OpenSSL 3.x, see: https://wiki.mozilla.org/Security/Server_Side_TLS.

Here is a PR working on this issue: https://github.com/Kong/kong/pull/12420

**Comment by zhao173205** [comment]
Dear team




Thank you for your reply. After testing, when I lowered the security level of OpenSSL's/etc/ssl/openssl.cnf to CipherString=DEFAULT: @ SECLEVEL=0, handshake communication can proceed normally







At 2024-01-26 11:42:30, "Niklaus Schen" ***@***.***> wrote:

It seems to be related to the OpenSSL version upgrade.

Kong3.1 uses OpenSSL 1.1.1s, while kong 3.4 uses OpenSSL 3.1.4. TLSv1.1 is not supported by default in OpenSSL 3.x, see: https://wiki.mozilla.org/Security/Server_Side_TLS.

Here is a PR working on this issue: #12420
